### DISCUSSION TITLE
Why Does `EXTCODECOPY` Always Throw an Error?

### DISCUSSION THREAD

**Discussion by idea404** [zkSync-Community-Hub/zksync-developers#86] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

I want to implement a contract method that fetches the bytecode of a deployed contract on another account. Currently the documentation accurately points out that this [always throws an error on zkEVM](https://era.zksync.io/docs/reference/architecture/differences-with-ethereum.html#extcodecopy). I was wondering however why is this? 

### What is the specific issue or error you're encountering?

The terminal output containing the error:

```bash
An unexpected error occurred:

Error: Command failed: /Users/dennis/Library/Caches/hardhat-nodejs/compilers-v2/zksolc/zksolc-v1.3.13 --standard-json  --system-mode  --solc /Users/dennis/Library/Caches/hardhat-nodejs/compilers-v2/wasm/soljson-v0.8.17+commit.8df45f5f.js.executable
Contract `contracts/Tutorials.sol:Tutorial` compiling error: The contract `contracts/Tutorials.sol:Tutorial` LLVM IR generator definition pass error: 49:25 The `EXTCODECOPY` instruction is not supportedThe contract `contracts/Tutorials.sol:Tutorial` LLVM IR generator definition pass error: 49:25 The `EXTCODECOPY` instruction is not supported


    at ChildProcess.exithandler (node:child_process:402:12)
    at ChildProcess.emit (node:events:513:28)
    at ChildProcess.emit (node:domain:489:12)
    at maybeClose (node:internal/child_process:1100:16)
    at Process.ChildProcess._handle.onexit (node:internal/child_process:304:5) {
  code: 1,
  killed: false,
  signal: null,
  cmd: '/Users/dennis/Library/Caches/hardhat-nodejs/compilers-v2/zksolc/zksolc-v1.3.13 --standard-json  --system-mode  --solc /Users/dennis/Library/Caches/hardhat-nodejs/compilers-v2/wasm/soljson-v0.8.17+commit.8df45f5f.js.executable'
}
error Command failed with exit code 1.
```

### Can you share the error messages or logs you're receiving, if any?

Shared above

### Have you made any recent changes to the contract before encountering this issue?

No

### Are there any external libraries or contracts that your contract interacts with?

No

### Can you provide the relevant portions of your contract code where the issue is occurring?

```solidity
// SPDX-License-Identifier: MIT
pragma solidity >=0.8.17;

contract Tutorial {
    function getCode(address _addr) public view returns (bytes memory o_code) {
        return _addr.code;
    }
}
```

### Have you tried to isolate the problem, and if so, what were the results?

NA

### What steps have you already taken to try to resolve the issue?

NA

**Comment by MexicanAce** [comment, marked as answer]
@idea404 Per the zkSync docs:
> Contract bytecode cannot be accessed in our architecture. Only its size is accessible with both CODESIZE and EXTCODESIZE.

https://era.zksync.io/docs/reference/architecture/differences-with-ethereum.html#extcodecopy

**Reply by idea404** [reply]
Maybe I phrased my question wrong: how does the zkEVM differ from EVM such that it disallows the use of `EXTCODECOPY`?

**Comment by bricnounlerumark** [comment]
The issue you're encountering is due to the fact that the EXTCODECOPY instruction is not supported on zkSync's zkEVM. This instruction is typically used to copy code from an external contract, which is not allowed in zkSync's environment. As a result, attempting to fetch the bytecode of a deployed contract on another account will always result in an error on zkEVM. You may need to explore alternative approaches to achieve your desired functionality within the constraints of zkSync's environment.

**Comment by platactiomeb** [comment]
Issue:
Attempting to fetch bytecode of a deployed contract on another account using the `EXTCODECOPY` instruction throws an error on zkEVM.

Reason:
The `EXTCODECOPY` instruction is not supported on zkEVM, resulting in a compilation error when attempting to use it in the contract code.

Resolution:
Avoid using `EXTCODECOPY` instruction or similar operations when fetching bytecode on zkEVM, as it is not supported. Consider alternative methods for achieving the desired functionality.

**Comment by sfatitetyccindy** [comment]
Issue: Contract method to fetch bytecode of deployed contract on another account always throws an error on zkEVM.
Reason: zkEVM does not support the EXTCODECOPY instruction, resulting in compilation errors.
Next Steps: Check zkEVM documentation or contact zkEVM developers for alternative methods to fetch bytecode.

**Comment by epjocnopasara** [comment]
The issue you're encountering with the EXTCODECOPY instruction not being supported on zkEVM is likely due to zkSync's design choices and limitations. zkEVM is a layer-2 Ethereum Virtual Machine (EVM) implementation that operates within zkSync's rollup architecture, aiming to provide scalability and security benefits. However, certain EVM opcodes and instructions may not be directly supported or feasible within zkEVM due to its design and optimization goals. As a result, attempting to execute operations like EXTCODECOPY may lead to compilation errors or runtime issues. For fetching bytecode of deployed contracts on another account in zkSync, alternative approaches or solutions may be needed due to these limitations.

**Comment by venlaudecsubs** [comment]
The user is trying to implement a contract method that fetches the bytecode of a deployed contract on another account on the zkSync testnet. However, the documentation accurately points out that this always throws an error on zkEVM and the user is wondering why this is the case. The user also encountered an error when compiling a contract, which is related to the EXTCODECOPY instruction not being supported on zkEVM. No specific error messages or logs are provided, and no recent changes have been made to the contract.
